Add framework for batch compression of shuffle partitions

jlowe opened the pull request:
This adds `TableCompressionCodec`, an interface for compression codecs that can compress contiguous table buffers.  A single trivial copy codec is provided just for testing.  Other codecs can be added in the future.

Batch compression is supported and utilized in the partitioning code when using the RAPIDS UCX shuffle plugin which produces contiguous tables.  Batch decompression is supported and utilized in the batch coalescing that is inserted into the query after each shuffle.  The coalesce code automatically detects compressed shuffle buffers and handles a mix of uncompressed and compressed batches arriving to be coalesced.

If desired I can split this PR into smaller pieces (e.g.: just adding TableCompressor and copy codec, then separate PR to plug it into partitioning and coalescing).  I opted to do a single PR initially so it was clear how the table compression APIs were being used in practice.

jlowe commented:
build

jlowe commented:
build

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:
```diff
@@ -140,7 +140,12 @@ class RapidsDeviceMemoryStore(
         GpuColumnVector.from(table.get) //REFCOUNT ++ of all columns
       } else {
-        throw new UnsupportedOperationException("compressed buffer support not implemented")
+        val uncompressedBuffer = uncompressBuffer(contigBuffer, meta.bufferMeta)
```
In practice `getColumnarBatch` needs to be called a lot less for compressed buffers. Maybe a note would be good to remind us to be careful when compression is on.

You could also change `getColumnarBatch` to be something like `getOrUncompressBatch` to make it abundantly clear.

I think the pattern will likely be: call this function and the remove the buffer, so this function should be getting called once.

jlowe replied:
Yes, it minimally adds a reference when it's called and must be separately closed for each `getColumnarBatch` call, so I don't think we're going to see code spamming this for the same columnar batch.  I will add a node to `RapidsBuffer.getColumnarBatch` to be explicit.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:
```diff
@@ -590,8 +590,27 @@ object RapidsConf {
     .createWithDefault(50 * 1024)
 
+  val SHUFFLE_COMPRESSION_ENABLED = conf("spark.rapids.shuffle.compression.enabled")
+      .doc("Whether to enable compression of shuffle buffers")
```
Can we make it clear here too that this is only for the RAPIDS shuffle.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+  /**
+   * Compress a contiguous table.
+   * @note The contiguous table is NOT closed by this operation and must be closed separately.
+   * @note The compressed buffer MAY NOT be ideally sized to the compressed data. It may be
```
Yes, this means you'll get an output buffer that is the original estimate of the output size.  The amount of compressed data within the buffer may be significantly less than the size of the buffer allocated to contain it.  It's up to the caller to decide if it's worth copying the data to a new buffer to release the wasted space.

Agree that it's obvious that the compressed data starts at 0.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+      }
+      closeOnExcept(tempBuffer) { tempBuffer =>
+        val outputSize = getOutputSpaceNeeded(contigTable.getBuffer, tempBuffer)
+        memNeededThisBuffer += outputSize
```
`memNeededToCompress`?

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+  }
+
+  private def compressBatch(): Unit = {
+    if (tables.nonEmpty) {
```
nit, could move this if to the `compressBatch` method definition, so spacing goes to the left.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+ *                           be compressed individually.
+ */
+abstract class BatchedTableCompressor(maxBatchMemorySize: Long) extends AutoCloseable with Arm {
+  private[this] val tables = new ArrayBuffer[ContiguousTable]
```
could we have comments for these? I am not entirely sure what `oversizedOutBuffers` is. I believe it is larger than the regular device buffer but a comment here would help.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+ *                           be compressed individually.
+ */
+abstract class BatchedBufferDecompressor(maxBatchMemorySize: Long) extends AutoCloseable with Arm {
+  private[this] val inputBuffers = new ArrayBuffer[DeviceMemoryBuffer]
```
same here might be nice to have a description of these buffers.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+  private[this] var batchMemUsed: Long = 0
+
+  /** The codec ID corresponding to this decompressor */
+  val codecId: Byte
```
so here we refer to `codecId` instead of `codecType` so still a nit, but would be nice to pick one.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+  /**
+   * This must be called after all buffers to be decompressed have been added to retrieve the
+   * decompression results.
+   * @return compressed tables
```
```suggestion
   * @return decompressed tables
```
I hope ;)

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+    }
+  }
+
+  /**
```
`s/decode/decompress` in this function.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+  /** The codec ID corresponding to this decompressor */
+  val codecId: Byte
+
+  def addBufferToDecode(buffer: DeviceMemoryBuffer, meta: BufferMeta): Unit = {
```
```suggestion
  def addBufferToDecompress(buffer: DeviceMemoryBuffer, meta: BufferMeta): Unit = {
```

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+   * batch compressor which is responsible for closing the table.
+   * @param contigTable the contiguous table to be compressed
+   */
+  def addTable(contigTable: ContiguousTable): Unit = {
```
nit, `addTableToCompress`

abellina commented:
@jlowe I got through the rest of it. Let me know when to re-check.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:
```diff
@@ -384,21 +386,75 @@ class GpuCoalesceIterator(iter: Iterator[ColumnarBatch],
+    val sizes = new Array[Long](numCols)
+    (0 until numCols).foreach { i =>
+      tableMeta.columnMetas(columnMeta, i)
+      var subBuffer = columnMeta.data(subBufferMetaObj)
```
This doesn't compile.  `columnMeta.data` has a both parameterless form and a single-parameter form, therefore Scala thinks it is a sequence of `SubBufferMeta` which is not what we want.  I get what you're trying to suggest here, but I'm not sure how to easily get Scala to reference the proper Java method because of the overloads.

abellina replied:
Thanks for trying, sorry I should have tried to compile this.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala:
```diff
@@ -0,0 +1,367 @@
+   * batch compressor which is responsible for closing the table.
+   * @param contigTable the contiguous table to be compressed
+   */
+  def addTable(contigTable: ContiguousTable): Unit = {
```
Yes, the batch compressor itself must be closed at some point.  Note that `addTable` takes ownership of the table up to the point where it is placed into the `tables` tracking array in the batch compressor.  At that point `addTable` is no longer responsible for it (hence the `closeOnExcept` clause used in that method instead of `withResource`), and the table is now owned by the batch compressor.  Once the batch compressor is closed, all tables in the `tables` array will be closed.

jlowe commented:
@revans2 @abellina I believe I've addressed all the review comments, so this is ready for another look.

jlowe commented:
build

jlowe commented:
This needs to be upmerged to the latest on branch-0.2.

jlowe commented:
@abellina @revans2 There were some tests that needed updating after the latest changes on branch-0.2.  I fixed handling degenerate batches and added tests for that case.  I also fixed a buffer leak I saw during manual testing that was occurring in partitioning.

jlowe commented:
build

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/MetaUtils.scala:
```diff
@@ -132,12 +132,18 @@ object MetaUtils extends Arm {
-  def buildDegenerateTableMeta(tableId: Int, batch: ColumnarBatch): TableMeta = {
+  def buildDegenerateTableMeta(batch: ColumnarBatch): TableMeta = {
     require(batch.numRows == 0 || batch.numCols == 0, "batch not degenerate")
+    if (batch.numCols > 0) {
```
I am not sure how a degenerate table can be compressed. Maybe it's a byproduct of all tables getting compressed. Perhaps a comment might help?

jlowe commented:
build

jlowe commented:
build

jlowe commented:
@revans2 can you take another look?  I needed to upmerge to the latest on branch-0.2.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:
```diff
@@ -590,8 +590,27 @@ object RapidsConf {
     .bytesConf(ByteUnit.BYTE)
     .createWithDefault(50 * 1024)
 
+  val SHUFFLE_COMPRESSION_ENABLED = conf("spark.rapids.shuffle.compression.enabled")
```
I was emulating Spark's shuffle config where the compression codec and compression enable are separate.

jlowe replied:
Updated this to use a single config for the shuffle codec, with `none` indicating no shuffle compression codec should be used.

jlowe commented:
Holding off on the CI build for now.  Waiting on the HashJoin fix for 3.1 and updated cudf with the fix for https://github.com/rapidsai/cudf/pull/5915 which was found during the added unit tests in this PR.

jlowe commented:
build
